src/combinatorics/composition.lean

```diff
@@ -381,31 +378,213 @@ begin
+    rw eq_cons_of_length_one A at B ⊢,
+    simpa [single_blocks] using B }
 end
 
 end composition
 
+/-!
+### Splitting a list
+
+Given a list of length `n` and a composition `c` of `n`, one can split `l` into `c.length` sublists
+of respective lengths `c.blocks_fun 0`, ..., `c.blocks_fun (c.length-1)`. This is inverse to the
+join operation.
+-/
+namespace list
+variable {α : Type*}
+
+/-- Given a list of length `n` and a composition `[i₁, ..., iₖ]` of `n`, split `l` into a list of
+`k` lists corresponding to the blocks of the composition, of respective lengths `i₁`, ..., `iₖ`.
+This makes sense mostly when `n = l.length`, but this is not necessary for the definition. -/
+def split_wrt_composition (l : list α) (c : composition n) : list (list α) :=
+of_fn (λ (i : fin c.length), (l.take (c.size_up_to (i.val+1))).drop (c.size_up_to i.val))
+
+/-- When one splits a list along a composition `c`, the number of sublists thus created is
+`c.length`. -/
+@[simp] lemma length_split_wrt_composition (l : list α) (c : composition n) :
+  length (l.split_wrt_composition c) = c.length :=
+by simp [split_wrt_composition]
+
+/-- When one splits a list along a composition `c`, the lengths of the sublists thus created are
+given by the block sizes in `c`. -/
+lemma map_length_split_wrt_composition (l : list α) (c : composition l.length) :
+  map length (l.split_wrt_composition c) = c.blocks :=
+begin
+  refine ext_le (by simp) (λ i h₁ h₂, _),
+  simp [split_wrt_composition, composition.size_up_to_le c, min_eq_left, nth_le_of_fn',
+             map_of_fn, function.comp_app, length_drop, length_take],
+  rw c.blocks_length at h₂,
+  simp only [composition.size_up_to_succ c h₂, nat.add_sub_cancel_left]
+end
+
+lemma length_pos_of_mem_split_wrt_composition {l l' : list α} {c : composition l.length}
+  (h : l' ∈ l.split_wrt_composition c) : 0 < length l' :=
+begin
+  have : l'.length ∈ (l.split_wrt_composition c).map list.length :=
+    list.mem_map_of_mem list.length h,
+  rw list.map_length_split_wrt_composition at this,
+  exact c.blocks_pos this
+end
+
+lemma sum_take_map_length_split_wrt_composition
+  (l : list α) (c : composition l.length) (i : ℕ) :
+  (((l.split_wrt_composition c).map length).take i).sum = c.size_up_to i :=
+by { congr, exact map_length_split_wrt_composition l c }
+
+/-- The `i`-th sublist in the splitting of a list `l` along a composition `c`, is the slice of `l`
+between the indices `c.size_up_to i` and `c.size_up_to (i+1)`, i.e., the indices in the `i`-th
+block of the composition. -/
+lemma nth_le_split_wrt_composition (l : list α) (c : composition n)
+  {i : ℕ} (hi : i < (l.split_wrt_composition c).length) :
+  nth_le (l.split_wrt_composition c) i hi = (l.take (c.size_up_to (i+1))).drop (c.size_up_to i) :=
+by simp [split_wrt_composition]
+
+/-- If one splits a list along a composition, and then joins the sublists, one gets back the
+original list. -/
+theorem join_split_wrt_composition (l : list α) (c : composition l.length) :
```

Comment:
The rest of this PR looks good to me. I wonder if there is some CS trick to compress this proof. @digama0 @cipher1024 do you have an idea?

Reply:
Yes, this proof is nonsense, I agree. But it works :)

Reply:
@sgouezel Let's make these two theorems `simp`.
Otherwise LGTM.

Reply:
I have added the simp attributes.